### ISSUE TITLE
Bookmark organization

### ISSUE THREAD

**Issue by sachali** [minbrowser/min#697]
I'm pretty sure this issue must have been raised, but I would like to add my vote for deeper bookmark functionalities: grouping bookmarks/folders, renaming, and either a bookmark tab on the manu bar or a dedicated bookmark page (like preferences). This is like the only thing that's keeping me from completely switching to Min.

**Comment by PalmerAL** [comment]
Do you mind if I ask what you typically use bookmarks for? In particular:

* How do you typically group bookmarks together - things related to a specific project? a site? etc.
  * Do you use nested folders?
* When you create a bookmark, is it typically because you're going to visit the page frequently, or because you want to come back to it at some point a long time in the future (and want it to be saved forever)?
* When you rename a bookmark, is it typically because the page title doesn't reflect the actual content of the page, or because you want to make a note about something else (why you saved the page, etc.)

I think some form of bookmark management would be nice, but I'm not sure what the best way to do it is. Existing browsers have really complicated ways of managing bookmarks (nested folders, multiple locations, smart folders, keywords, tags), and I'd like to avoid adding all of that, but I'm not sure which features people actually use. Additionally, I think it would be nice if it was easier to manage bookmarks - instead of having to categorize every bookmark yourself, the browser could automatically organize everything for you and help you find what you're looking for. However, automatic organization is never going to work 100% of the time, so it would still need to be easy to ignore that and organize your bookmarks yourself (or maybe I'm thinking about this wrong, and bookmark organization is something that's complicated enough it needs to be done manually).

Regarding the bookmark page, I'm curious what you think the benefit of that would be over what we already have (view menu > bookmarks)? It seems like we could add some basic folder/editing options to that without it being too overwhelming, but maybe not.

**Comment by sachali** [comment]
I’m writing with the awareness that Min is a different browser with different goals from the ones I’ve used previously, and so shouldn’t expect it to have all the features I enjoyed while maintaining its minimalism. That being said, I migrated from Safari and have dabbled in Chrome, so how I do bookmarks was probably highly influenced by how bookmarks are done in these browsers.

> How do you typically group bookmarks together - things related to a specific project? a site? etc.

Usually I group bookmarks based on topics/interests. So I have folders for literary magazines, academic journals, research tools, (things related to) medical humanities, etc.
If you count the bookmark bar as one way of sorting that that would be sorted by frequency.
I never thought of bookmarking a page I know I only need for this specific project and never after, but I might start doing that and delete the whole folder after the project is done.

> Do you use nested folders?
I use nested folders (although most of my folders are only one level deep, and I’d be ok if there’s no nested folders), but I don’t mind using nested tags the way Bear does it.

> When you create a bookmark, is it typically because you're going to visit the page frequently, or because you want to come back to it at some point a long time in the future (and want it to be saved forever)?

Both… I guess?

On Safari’s bookmark bar I have my frequently visited pages (which also appear on a blank new tab—so maybe this is something you can consider since right now a new tab on Min is just blank) and some scripts (save to Instapaper)

I’m an academic and translator, so most of the bookmarks are academic resources and translation tools. They are things that I want to come back to in the future in the sense that I am temporarily out of school and currently not working on a specific project and so wouldn’t be using them. But when I am I’ll use them frequently. Just to be clear these are not articles but proper websites as my goal is try not to make my bookmarks a read-it-later dump site.

> When you rename a bookmark, is it typically because the page title doesn't reflect the actual content of the page, or because you want to make a note about something else (why you saved the page, etc.)

The former. I want my bookmarks to be clear and easily comprehensible. When I rename a bookmark it’s usually to delete the unnecessary or redundant parts (the “subtitles”?) in the name. For example, if I bookmark [Min](https://minbrowser.github.io/min/), the name of this bookmark will be “Min | A smarter, faster web browser,” and I’d like this page to simply be named “Min.” Some pages are far worse in that their titles are some sort of scripts or begin with special characters or irrelevant information and don’t even include their actual name.
This is especially important on Safari to save real estate on the bookmark bar, where I have my most visited pages, facebook, youtube, google docs, word online, etc. So “Microsoft Word Online - Work Together on Word Documents” is simply “Word,” and for youtube it’s ⏯.

> Additionally, I think it would be nice if it was easier to manage bookmarks - instead of having to categorize every bookmark yourself, the browser could automatically organize everything for you and help you find what you're looking for. However, automatic organization is never going to work 100% of the time, so it would still need to be easy to ignore that and organize your bookmarks yourself (or maybe I'm thinking about this wrong, and bookmark organization is something that's complicated enough it needs to be done manually).

Personally, I’m not sure if automatic sorting is going to be very useful for my use cases. 

I have folders for different academic disciplines (anthropology, sociology, philosophy), but I also have another dedicated folder for medical humanities/, which is my academic focus. And my system is that this folder takes priority over the other ones, meaning it contains medical anthropology, sociology, philosophy, and even hospital websites. This is probably peculiar and most people don’t organize their bookmarks this way, so it would be very interesting to learn about their organization patterns.

> Regarding the bookmark page, I'm curious what you think the benefit of that would be over what we already have (view menu > bookmarks)? It seems like we could add some basic folder/editing options to that without it being too overwhelming, but maybe not.

Honestly I wasn’t aware of view > bookmarks until now lol. I have no problem with the way it is now as long as organizing and editing can easily be done.

**Comment by PalmerAL** [comment]
Thanks for providing such a detailed explanation, that's really helpful!

Regarding a couple specific things you mentioned:
> I want my bookmarks to be clear and easily comprehensible. When I rename a bookmark it’s usually to delete the unnecessary or redundant parts (the “subtitles”?) in the name

That's a good point, and I can see how some way of supporting that kind of naming would be useful. We actually have some code right now that attempts to simplify titles automatically in the search results (for example, the "| ars technica" text gets removed from the title of [this page](https://arstechnica.com/gadgets/2019/01/next-windows-10-version-will-let-you-search-without-cortanas-involvement/)). In addition to manual naming, we could probably make some changes to make that work in more situations, and start doing that for the !bookmarks view as well.

> Honestly I wasn’t aware of view > bookmarks until now lol. I have no problem with the way it is now as long as organizing and editing can easily be done.

You're not the only one! A lot of people don't seem to realize that that option exists, and I'm not sure what to do about it. I moved it closer to the top of the menu in 8595ccb41477bf95abcd2f8b4460e46fe86ee9ed so it should be slightly easier to find, but if you have any other ideas about this I'd be interested to hear them.

> right now a new tab on Min is just blank

I'm not sure an entire new-tab page is a good idea, but we could consider changing the suggestions that show up in a new tab. Right now, you should be seeing 4 frequently-visited sites - we could change that to rank bookmarked sites higher so that those get suggested more often, or increase the number of results that get shown.

In general, tags sound like a pretty good solution to organization, particularly for the part you mentioned with an overall "medical" folder. Compared to folders, tags should also have a simpler UI, because you don't have to represent a tree structure, so I'm inclined to do that rather than have actual folders unless there's a strong objection to it.

There's a couple of options that I could see working for this:
* Add a sidebar to the task overlay with access to bookmarks (either a single "bookmarks" item, or one menu item per folder). So the top item (which would be visible when you open the overlay) would be things you have "open now", and the bookmarks items would be "things from the past".
  * This would help make the bookmarks menu easier to find, but it also makes it more confusing what the task overlay actually does.

* Add a bar along the top of the existing bookmarks view that allows you to select tags to filter by.
  * With this option, bookmarks are more hidden, and it takes more clicks to access them, but they also shouldn't get in your way as much when you don't want them.
  * It might also be kind of overwhelming to show so many options at once.

I'll try to make some mockups of this soon in case it's not clear how this would work from my description.

Or maybe neither of these is best - is there another place to put bookmarks that you think would work well?

**Comment by Pat38** [comment]
Hi PalmerAL,
I'm new to github. Just writing to say that you are doing a great job and that it would be nice to have a left sided panel for the bookmarks, next to the current tab, as in Safari. I'm almost ready to dump Firefox for your web browser but I need desperately a good bookmarks manager with rename option, import/export, folders and subfolders. 
Thank you very much 👍
Patrick

**Comment by PalmerAL** [comment]
@Pat38 Thanks! I'm going to try to address each thing you mentioned separately:

* Renaming makes sense, I think we are going to implement that at some point.
* Import/export would also be nice to have, although it's fairly complicated, and I'm really sure where it would fit in the UI.
* It seems to me like tags are generally more flexible and easier to use than folders - you don't have to organize your bookmarks in a linear category > subcategory structure, but you can choose to do so if you want (by having a tag for the outer folder and separate tags for each inner folder). Based on how you use bookmarks, do you think this would work, or is there another advantage to folders that I'm not seeing?
* I'm not sure I really understand the usecase for a side panel - can you explain more? Do you have a few bookmarks that you access really frequently (but too many to fit in the regular bookmarks bar)? Or do you have a widescreen monitor so it's better to use horizontal space rather than vertical? Something else?

**Comment by PalmerAL** [comment]
@sachali I made a mockup of the "Add a bar along the top of the existing bookmarks view" concept, which I think might work well for this.

When you open the bookmarks view, you'll see a list of tags along the top, sorted from left-to-right based on a combination of how frequently you use those tags, and how recently you've added a bookmark with that tag:

<img width="1440" src="https://user-images.githubusercontent.com/10314059/53292372-0970a280-3787-11e9-802b-27acd838a12a.png">

When you select a tag (either by clicking or by typing an @ and the tag name in the search bar), the results in the list will change to show only bookmarks with that tag, and the tag will show up as selected. Additionally, a series of other tags will appear to the right of it. These tags will be sorted based on how frequently you use them in combination with the tag you've already selected, so if you have a bunch of bookmarks tagged with "medicine" along with a specific subcategory, those subcategories will show up after you select the "medicine" tag:

<img width="1431" src="https://user-images.githubusercontent.com/10314059/53292393-610f0e00-3787-11e9-8dac-01f373e0751d.png">

Next to each bookmark in the list, there would be a button you could click to show a dropdown, which would allow you to add or remove tags.

What do you think?

**Comment by MarkWahlsten** [comment]
@PalmerAL, just want to add a strong upvote to using a tagging system for organising bookmarks! 👍  So many times I've needed to save things in multiple categories. 

Is it worth seeing how it's been implemented in some writing/note apps like Bear (as sachali pointed out), Evernote, Google Keep... or productivity apps like Wunderlist, Asana, Trello... as they've had time to iterate on it and possibly figure out what does or doesn't work?

I'm pretty happy with what you've outlined and mocked-up already, but just in case they've already solved problems that could be applicable to Min. 

**I'm a little bit "massive long-term dump and hoard" and a little bit "one-off projects via Pinterest/Trello", with a dash of "frequently used on the toolbar" ... all while quietly wishing I could commit to maybe just one of them**

_(But that would mean facing the giant task of consolidating one way or the other – it's gotten to Reorganise My Garage size now, it's too late for it now, abandon ship!)_

- To a degree I've had trouble settling on a method of saving bookmarks: my storage often becomes so overwhelmingly bloated and disorganised I abandoned most of it in favour of the bookmark bar.
- This is partly for above "linear folder" reason (I could never remember where I put anything), but mostly because ADD. I hoard tabs because I can't decide whether or not I need them, so to cull them down I sidestep the decision altogether and pass it to Future Me. I fire off pretty much any tab that isn't nailed down to Pinterest, or they stay in the forever in the browser, like fungi. 
- The kinds of things I hoard can be a mix of "apps/resources that I don't know if I need yet, but sound like something I may use/reuse", "inspiration (either for something specific or a great repository) I don't know if I need yet, that I may use/reuse", and "articles that I don't know if I need to read yet, but would need to read them to find out, and once I've done that I don't need the link anymore (if I have no plans to share it)"
- I tried "read later" lists for the articles, but I'd forget about them and end up with the same issue as with bookmarks (disorganised and huge), so it didn't make much sense to separate them from all the other links (but I dunno may I was just doing it wrong)
- The other issue was swapping back/forth between browsers: each had their own set of bookmarks, nothing was synced 

**The above reasons is why I ended trying to shift to Pinterest or Trello: to keep everything in one place** 
- I'm more of visually-oriented person, I get overwhelmed pretty easily by giant lists of text without visual context to quickly recall what they were for (and break up the wall of text a bit), which is probably why I've abandoned so many bookmark lists. In this respect the card/mood board layout make sense for me (but maybe not necessarily for Min in general). 
- As mentioned above, I also use Pinterest/Trello to group things by project. I do a lot of visual research in Pinterest anyway, so it makes sense to save relevant non-inspiration links with the rest 
- Which app (and which browser) I used depended largely on compatibility – I used to favour Pinterest until its extensions kept getting blocked by ad blockers, but now they favour bookmarklets, so that side-stepped extension compatability altogether.* 

Thanks a bunch! Love Min, keep it up! :)

_* @PalmerAL – I talked at length in my survey response about bookmarklets, but have since seen your reference to [userscripts](https://github.com/minbrowser/min/wiki/userscripts). Do these work in a similar way to bookmarklets, would they enable use of Pinterest and/or Trello? (Is there an existing thread or resource you could point me to if this or a similar use case has already been covered?)_

**Comment by PalmerAL** [comment]
@MarkWahlsten Thanks for the ideas! Sorry for the delay, it's taken me a couple of days to have time to respond to this.

I haven't used most of those apps before, so it's hard to say for sure, although none of them look all that differerent from what I proposed above. Several of them ([bear](https://bear.app/faq/Tags%20&%20Linking/Nested%20Tags/), [wunderlist](https://6wunderkinder.desk.com/customer/portal/articles/1641444)) seem to be very focused on making it easy to tag things, which I guess makes sense considering that it easy to forget. I'm not really sure how the actual tagging process should work, I suppose that's something to consider.

You description of  bookmarks being a giant list of things-I-might-want-to-look-at-someday matches my experience pretty closely; my hope is that we can find a way to make organizing bookmarks really easy, so that you end up with something other than a huge list, but I don't know if tags alone are enough to solve that.

I haven't used Pinterest before - how do the images of pages work? Is is the entire page, or does it auto-select an image from the page, or do you pick an image? A while ago, Firefox had screenshots of whole pages on the new tab page, and it never seemed to me like they were that useful for figuring out what a page was, but maybe it is helpful for other people.

I tend to agree with you that a reading list isn't much better than a normal bookmarks list, and I've been thinking about removing the reading list in Min and just making it easier to bookmark articles instead; having a special "articles from reader mode" bookmark tag that that gets automatically applied might be a good way to solve this.

Syncing probably won't happen anytime soon, both because it's a pretty complex project and would require a decent amount of money to run, but also because it wouldn't be useful to most people unless we also had a mobile version (which is a huge project of its own... I think making one browser UI is enough for now). I've also thought about integrating with existing cloud storage providers (dropbox, google drive, etc) for sync, but most of those are designed for files, and I'm not sure how we would sync a database-based structure with that.

Userscripts aren't exactly like bookmarklets (userscripts run JS in the background whenever a page is loaded, bookmarklets run JS whenever you click on the bookmark), but you definitely can use pinterest inside a userscript. Here's a script you can use that will show the Pinterest UI if you press ctrl+e: https://gist.github.com/PalmerAL/7b29c45ba80391f76a636f8846e01ce5.

**Comment by samxiu** [comment]
the bookmarks features are good ideas. I'm using several browsers. Some browsers have features to import/export the bookmarks to other browsers. The bookmark features are working good enough but yet it is not perfect. Nowsaday, the security and privacy things on browsers are pretty good but sometimes it just to much work and hassle. example: google chrome, if you use it on not your devices, you need to go through the process of verification. I find out many people log in to theirs account and most of them forget to log out. Right now the easy way for myself, I'm using Cloud Server call Mega (there are many Free cloud Server you can use such as Google drive, dropbox, box, icloud, etc). I create a an excel spread sheet where all my bookmarks are atore in the Mega. So. I don't need to log in my google account to get and share the bookmarks. I just go to Mega and grab the link of the bookmars. I can use any browsers, any devices that not belong to me to get my bookmarks. And in the Mega, I set it up to auto upload the bookmarks. Using an excel spread, I can organize, sort, and search the bookmarks link. I like to use Min browsers and I'm hoping that Min have the bookmark features soon

**Comment by PalmerAL** [comment]
I'm going to do this in the next release.

**Comment by MarkWahlsten** [comment]
Hi @PalmerAL, 

Apologies I'm a bit sporadic with my replies (mostly because I ramble at such great length, it's all I know!) 

After seeing the 1.12 beta pop up, I was really keen to try out the bookmark tags. 

**I love the tags!** 

- Searching bookmarks makes so much more sense now. 
- My favourite part is how you can toggle the tags on/off without having to retype/delete anything 
- I haven't yet had a chance to test the "tags most used togther will appear together in searches" feature you described above


**My notes** - _feel free to pick out what's useful and ignore whatever's irrelevant_ 

- I admit, up until coming back to this thread after the 1.12 release, I had forgotten the !bookmarks page existed, despite it being available right there in the View menu (d'oh). 
    - I guess I mostly avoided the bookmark list for the reasons I listed in my previous post. 
    - (I also only use min in fullscreen, so I forgot there even was a View menu, haha)
    - Another reason may be – since I've been able to open/collapse Tasks (thankyou so much) I had started using Tasks as a kind of way to save/group bookmarks under specific interests, clients or projects (or as a dumping ground for links I've been hoarding that I can't get rid of).
    - That possibly isn't the best way to do things, memory-wise (there are a LOT of tabs)
    - Once I've got in the swing of tagging, I can hopefully cull all those Task groups back and return to using them as just a few workspaces for different clients/email accounts

- I wasn't really sure where to go to add the tags initially
    - My first instinct was to try and add a tag from within the url bar (after clicking on the bookmark star)
    - It was only after coming back to this thread that I remembered the !bookmarks page existed
    - If I hadn't been specifically looking for the tagging feature, I might not have known it was there unless I stumbled onto it by accident while editing a bookmark

**What were your plans for tags in future releases?**

- Are there plans to include the ability to add tags as you're bookmarking a page? 
- My initial thoughts on the interface was a second tags bar appearing below the URL bar (as per the !bookmarks page, when editing a bookmark's tags), the bar could just be the hollow "add tag" button, followed by most recent tags (or however they're organised)
    - ...or possibly much further down the track... tag suggestions? (whether existing or new tags?)
    - Barely an idea, I get the feeling that would be very complex (and possibly require AI), so feel free to ignore it if that kind of thing is way out scope
- I suggest this feature mainly because at the moment, a barrier for me properly using tags is the prospect of going back through and tagging all my old links 
    - ...and remembering to regularly go back and add tags to any new links I save
    - I'm highly skeptical I will remember to do this, haha

- Given this is a beta release, I get you're likely just working on getting the whole tagging system working behind the sceness first (and no doubt probably working on / thinking about most of this stuf anyway, I'm just doing a brain dump)

I really dig Min! 

Thanks again for all the hard work

**Comment by MarkWahlsten** [comment]
Hi @PalmerAL – in answer to your earlier questions (if any are still relevant)

> I haven't used Pinterest before - how do the images of pages work? Is is the entire page, or does it auto-select an image from the page, or do you pick an image? A while ago, Firefox had screenshots of whole pages on the new tab page, and it never seemed to me like they were that useful for figuring out what a page was, but maybe it is helpful for other people.

From what I can tell, it automatically pulls a selection of images from the page and lets you choose:

![Screen Shot 2019-12-18 at 1 00 25 pm](https://user-images.githubusercontent.com/20477121/71057544-04f2fb80-2198-11ea-9273-571c7446a7b1.png)

Hover displays 

![Screen Shot 2019-12-18 at 1 00 57 pm](https://user-images.githubusercontent.com/20477121/71057545-058b9200-2198-11ea-9a2d-f9d230c8aa8e.png)

Creating a new board (functions kind of like a folder):

![Screen Shot 2019-12-18 at 1 02 27 pm](https://user-images.githubusercontent.com/20477121/71057546-058b9200-2198-11ea-83eb-1b1e8c58c00d.png)

Here's what it does when it can't find any images:

![Screen Shot 2019-12-18 at 1 06 13 pm](https://user-images.githubusercontent.com/20477121/71058408-b72bc280-219a-11ea-8e64-01aa1ec4a4dd.png)

![Screen Shot 2019-12-18 at 1 03 58 pm](https://user-images.githubusercontent.com/20477121/71058467-e8a48e00-219a-11ea-98ba-c9cf3c8668b0.png)

![Screen Shot 2019-12-18 at 1 04 22 pm](https://user-images.githubusercontent.com/20477121/71058487-f0fcc900-219a-11ea-9963-6df988e95d44.png)

How they look on the Userscripts board:

![Screen Shot 2019-12-18 at 1 04 54 pm](https://user-images.githubusercontent.com/20477121/71058526-18ec2c80-219b-11ea-97a5-4968b244c8e2.png)


It's hard to say if I found Firefox's screenshots implementation very useful either, unless the page layout/colour was immediately recogniseable (e.g. like facebook); Safari does the same thing. Not really useful if it's just a page of text (like an article), a simple symbol or image was usually much less cognitive load. 

I've noticed Firefox's new tab page currently does both – logo (not sure where it's sourced) with favicon corner label, or a screenshot if there is no logo available:

![Screen Shot 2019-12-18 at 1 22 11 pm](https://user-images.githubusercontent.com/20477121/71058232-08878200-219a-11ea-81aa-7f99df9e2328.png)

I don't really use Firefox often, but I could find that useful (but that's just me, I guess everyone processes information differently). 

> Syncing probably won't happen anytime soon

Oh agreed definitely! I don't think syncing or the pinterest UI necessarily needs to be incorporated into Min – I was more highlighting that the links I saved to Pinterest being browser-independant was just the main reason why I found it a useful alternative to browers' built-in bookmarks lists. 

Also thankyou for the pinterest userscript!

Cheers

**Comment by PalmerAL** [comment]
I'm glad it's working well for you (and thanks for all the feedback)!

> Are there plans to include the ability to add tags as you're bookmarking a page?

There's an !addbookmark command that will create the bookmark and add tags (like !addbookmark tag1 tag2), but it's fairly hard to find. I like the idea of showing the tagging bar after clicking the star much better; I'll probably try to do that soon.

> ...or possibly much further down the track... tag suggestions? (whether existing or new tags?)

This actually exists! Once you start tagging more pages, you should see suggested tags show up with a dotted outline like this: https://user-images.githubusercontent.com/10314059/69779659-5e09d980-116e-11ea-9996-099cd3769288.png

That being said, the quality of the suggestions isn't great so far (although they seem to be accurate a decent percentage of the time). I'll probably look at improving this at some point, but if you have examples of where the suggestions don't appear or are wrong, let me know and I can look at it.

> I suggest this feature mainly because at the moment, a barrier for me properly using tags is the prospect of going back through and tagging all my old links

This is what the "similar items" section is for: https://user-images.githubusercontent.com/10314059/69779719-95788600-116e-11ea-997b-63936dd96387.png. Once you start tagging some of your bookmarks, related ones should show up in the list as well so you don't have to manually tag them. (Although once again, this depends on the tag suggestions being accurate, which may not be the case).

**Comment by MarkWahlsten** [comment]
No worries @PalmerAL

> There's an !addbookmark command that will create the bookmark and add tags (like !addbookmark tag1 tag2), but it's fairly hard to find. I like the idea of showing the tagging bar after clicking the star much better; I'll probably try to do that soon.

**Oh cool, I like that command. I just had a try of it, made some notes below.*** 

- What if you combined those two methods, so if you clicked the star it autofills the !addbookmark command (showing the tagging bar with suggestions underneath)? 
- Then if you already knew the !addbookmark command you could type away – now that I know it, I can definitely see it being a timesaver
    - Mostly I just like the idea of having a choice
- Clicking the star could possibly act as a visual prompt or reminder that the command exists?
    - I also wondered whether some of the less obvious shortcuts/commands could be pointed to via a sort of dismissable tooltip bar that appears under the address bar
        - or a line of text that sits above anything else (like url suggestions or search results)
    - ...or any other unintrusive contextual tip?
    - There's a lot of really great features and shortcuts in Min that have sailed right past me, I just wonder if there's a way to highlight them all without cluttering up the UI 
        - ...or overloading with too much info


***Notes on !addbookmark command**

![Screen Shot 2019-12-19 at 2 49 36 pm](https://user-images.githubusercontent.com/20477121/71151571-0511fa80-226f-11ea-9dc0-5d0cf2e93406.png)

- I instinctually typed the tags as "#tag1 #tag2",  
    - Haha I've now ended up with a tag that appears on the !bookmarks page as "##tag1 #tag2"
    - (is there a way to delete tags, like kind of like the below screen grabs from Google Keep?)
- Leaving the hashtags out defintely makes typing quicker, but it might be confusing if the tags then appear with hashtags on the !bookmarks page
    - Ah ok I tried with spaces and it still saves as a single tag, after trial and error I've realised the tag separator is a comma
- could be worth doing another survey on how ppl use tags / expect them to work...?


**The main parts I like in the below Google Keep example:**
- if you make a mistake: there's two ways to delete the tag as you type
- visual feedback: the underline under what you're typing and the addition of the pillbox-outline tabs below is a good indicator that the tag was saved (or how it turned out)
    - e.g. in Min, as you type tags after !addbookmark and press SPACE (or comma), it could be pushed to a new the "tag bar" below (and either underline what you just typed like Keep does, or clear the text field for the next tag
- rollover of the underlined text acts a link or a means to delete it
    - the link could act as a shortcut to the !bookmark page, autofilling that tag?
- rollover of the pillboxes reveals an "X" to delete the tag

![Screen Shot 2019-12-19 at 2 46 05 pm](https://user-images.githubusercontent.com/20477121/71151267-54a3f680-226e-11ea-8676-dd716bc86133.png)

![Screen Shot 2019-12-19 at 1 55 31 pm](https://user-images.githubusercontent.com/20477121/71150471-8025e180-226c-11ea-843b-94ac53e593ca.png)

![Screen Shot 2019-12-19 at 1 55 52 pm](https://user-images.githubusercontent.com/20477121/71150472-8025e180-226c-11ea-8179-f938c3b6d0a0.png)

**Comment by PalmerAL** [comment]
The tag bar now shows up when you click the star button (in 6c4c2c1b5397a02c6023a1864a754af9a7ea92db):

<img width="1440" alt="Screen Shot 2019-12-24 at 4 10 19 PM" src="https://user-images.githubusercontent.com/10314059/71425925-42f49180-2668-11ea-9777-0264df2ddc47.png">

I decided not to display !addbookmark in the searchbar because it makes it look like you then have to press enter to save the bookmark, which you don't.

Having more visual feedback for !addbookmark would be cool, but it'd be fairly complex to do, and I don't think it's worth it. Clicking the star is going to be the primary way to create a bookmark, so !addbookmark is really for if you create a lot of bookmarks and have the syntax memorized already.

> There's a lot of really great features and shortcuts in Min that have sailed right past me, I just wonder if there's a way to highlight them all without cluttering up the UI  

Yeah, I don't know the answer to that either. Tooltips are nice, but it's really easy to end up with too many, and I'm not sure which features are actually worth pointing out.

There isn't a way to delete tags right now; I think there should be, but I'm not sure where to put it.

**Comment by PalmerAL** [comment]
I've also changed !addbookmark to split tags based on whitespace and ignore hashtags: e652d717b698c9f59446a469ecc2a60c84c8224f

**Comment by sjehuda** [comment]
These resources might be relevant.

A solution to privately synchronize bookmarks using an XMPP account: https://github.com/minbrowser/min/discussions/2537

A joint effort to standartize bookmarks: https://github.com/jarun/buku/discussions/795
